### PR TITLE
Add consent enforcer service and consent executor

### PR DESCRIPTION
**Pull request by ThaminduDilshan**
### Purpose
This pull request introduces a new consent enforcer service to the authentication flow and integrates it with the backend service registry. Additionally a new consent executor in introduced to perform runtime consent evaluations/ collections during auth journeys.

### Approach
Integration of consent enforcer service:

* Added the `ConsentEnforcerServiceInterface` and the default service implementation that allows to evaluate and record user consents during runtime authentication flows

Flow execution enhancements:

* A new consent executor is added and registered in the executor registry that allows to evaluate and record user consents during runtime flow executions
* Improved auth assert executor to filter out only consented attributes when constructing the auth assertion

Consent management client enhancements:

* Implemented the `updateConsent` method in the consent service and default client implementation to allow updating existing consent records, and updated DTO conversions to use string representations for enum types (`backend/internal/consent/default_client.go`).

### Related Issues
- https://github.com/asgardeo/thunder/issues/1280

### Related PRs
- N/A

### Checklist
- [ ] Followed the contribution guidelines.
- [ ] Manual test round performed and verified.
- [ ] Documentation provided. (Add links if there are any)
    - [ ] Ran Vale and fixed all errors and warnings
- [ ] Tests provided. (Add links if there are any)
    - [ ] Unit Tests
    - [ ] Integration Tests
- [ ] Breaking changes. (Fill if applicable)
    - [ ] Breaking changes section filled.
    - [ ] `breaking change` label added.

### Security checks
- [ ] Followed secure coding standards in [WSO2 Secure Coding Guidelines](https://security.docs.wso2.com/en/latest/security-guidelines/secure-engineering-guidelines/secure-coding-guidlines/introduction/)
- [ ] Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.

## Summary by CodeRabbit

* **New Features**
  * Consent prompts during authentication; users can approve/deny specific data-purpose elements.
  * Consent recording and update flows with session-backed prompts and expiry handling.
  * Attribute filtering so only consented attributes are used downstream.

* **Bug Fixes**
  * More specific, user-facing error messages for consent operations and invalid requests.

* **Tests**
  * Broad test coverage added for consent resolution, recording, prompt expiry, and edge cases.

### CODE REVIEW

File: backend/internal/authn/consent/error_constants.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+			Key:          "error.consentenforcerservice.consent_search_failed_description",
+			DefaultValue: "Error while searching for consent records from the consent service",
+		},
+	}
+
+	// ErrorConsentUpdateFailed is returned when the consent service rejects the
+	// request to update an existing consent record with a client error.
+	ErrorConsentUpdateFailed = serviceerror.I18nServiceError{
+		Type: serviceerror.ClientErrorType,
+		Code: "AUTH-CES-1003",
```
Comment (rajithacharith) [inline backend/internal/authn/consent/error_constants.go:62]: We have followed `XXX-0000` error code pattern, Shall we use that

File: backend/internal/authn/consent/error_constants.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+import (
+	"github.com/asgardeo/thunder/internal/system/error/serviceerror"
+	"github.com/asgardeo/thunder/internal/system/i18n/core"
+)
+
+// Client-facing service errors for the consent enforcer service.
+var (
+	// ErrorConsentPurposeFetchFailed is returned when the consent service rejects the
+	// request to list consent purposes with a client error.
+	ErrorConsentPurposeFetchFailed = serviceerror.I18nServiceError{
```
Comment (rajithacharith) [inline backend/internal/authn/consent/error_constants.go:30]: Is this intended to be a I18nServiceError?

File: backend/internal/authn/consent/service.go [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+// buildPurposePrompts builds the list of ConsentPurposePrompt for the given purposes, applying attribute filtering,
+// user profile filtering, and consented element filtering.
+func buildPurposePrompts(purposes []consent.ConsentPurpose, requiredAttributes []string,
+	consentedElements map[string]bool, userAttributeSet map[string]bool) []ConsentPurposePrompt {
+	// For each purpose, determine which required elements still need consent
+	var promptPurposes []ConsentPurposePrompt
+	for _, purpose := range purposes {
+		var essential, optional []string
+		for _, elem := range purpose.Elements {
+			// Stage 1 – Attribute scope filter
```
Comment (rajithacharith) [inline backend/internal/authn/consent/service.go]: Do we need these Stage 1, stage 2 comments?

File: backend/internal/authn/consent/error_constants.go (hunk 1 above)
Comment (ThaminduDilshan) [reply backend/internal/authn/consent/error_constants.go:62]: All authn internal services follows this format. We might have to revisit all of them or revisit our policy for error code prefixes.

File: backend/internal/authn/consent/error_constants.go (hunk 2 above)
Comment (ThaminduDilshan) [reply backend/internal/authn/consent/error_constants.go:30]: Yes. We have deprecated `ServiceError` and should use `I18nServiceError` for all future services, and should migrate all existing implementations.

File: backend/internal/authn/consent/error_constants.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		Code: "AUTH-CES-1004",
+		Error: core.I18nMessage{
+			Key:          "error.consentenforcerservice.consent_create_failed",
+			DefaultValue: "Failed to create consent record",
+		},
+		ErrorDescription: core.I18nMessage{
+			Key:          "error.consentenforcerservice.consent_create_failed_description",
+			DefaultValue: "Error while creating consent record in the consent service",
+		},
+	}
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/authn/consent/error_constants.go:86]: These are returned if the downstream consent service return a client error. Reason for implement like this is, when we have atomic API support for consent, there can be client error scenarios. For the flow execution, this will always be a server error and it's handled in the consent executor

File: backend/internal/flow/common/constants.go [hunk 5]
```diff
@@ -166,6 +176,8 @@ const (
 	InputTypeOTP = "OTP_INPUT"
 	// InputTypePhone represents a phone number input type.
 	InputTypePhone = "PHONE_INPUT"
+	// InputTypeConsent represents a consent decisions input type.
+	InputTypeConsent = "CONSENT_INPUT"
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/common/constants.go:182]: But this is only attribute keys. I.e. attribute names. It doesn't contain attribute values right? So they're not PII

File: backend/internal/flow/executor/consent_executor.go [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+	appID := ctx.AppID
+	userID := ctx.AuthenticatedUser.UserID
+
+	if !e.HasRequiredInputs(ctx, execResp) {
+		logger.Debug("Required consent decisions not provided; checking if consent is needed")
+		return e.checkConsent(ctx, execResp, ouID, appID, userID)
+	}
+
+	logger.Debug("Consent decisions provided; processing consent decisions")
+	return e.handleConsentDecisions(ctx, execResp, ouID, appID, userID)
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/consent_executor.go:117]: We don't have support for configuring ou for consent yet as applications are not bound to a ou. Added a TODO comment to track this

File: backend/internal/flow/executor/consent_executor.go [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+	promptData, svcErr := e.consentEnforcer.ResolveConsent(
+		ctx.Context, ouID, appID, userID,
+		userAttributes, ctx.AuthenticatedUser.AvailableAttributes,
+	)
+	if svcErr != nil {
+		if svcErr.Type == serviceerror.ClientErrorType {
+			logger.Debug("Client error while resolving user consent", log.Any("error", svcErr))
+			execResp.Status = common.ExecFailure
+			execResp.FailureReason = "Failed to resolve consent: " + svcErr.ErrorDescription.DefaultValue
+			return execResp, nil
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/consent_executor.go:146]: But this is for a client error. Client needs to know why this request fails and hence we need to add these info to the response

File: backend/internal/flow/executor/consent_executor.go [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+
+		logger.Error("Failed to resolve consent", log.Any("error", svcErr))
+		return nil, errors.New("failed to resolve consent")
+	}
+
+	// All consents are active — nothing to prompt
+	if promptData == nil {
+		logger.Debug("All required consents are active; completing consent executor")
+		execResp.Status = common.ExecComplete
+		return execResp, nil
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/consent_executor.go:157]: No fix needed. The [promptData == nil] path is correct as-is. Here's why:

When [promptData == nil], ResolveConsent determined that all required consents are already active — the user previously approved everything the app needs. In this path:

- No new consent record is created, so there's no consent ID to set. Setting a stale/fake [RuntimeKeyConsentID] would be misleading.
- No filtering is needed in auth assert, because the user already consented to all required attributes. Not setting means the auth assert executor treats it as "consent was not a blocking factor" and includes all configured attributes — which is correct since the user already consented to them.

File: backend/internal/flow/core/prompt_node.go [hunk 9]
```diff
@@ -182,12 +182,7 @@ func (n *promptNode) hasRequiredInputs(ctx *NodeContext, nodeResp *common.NodeRe
 	}
 
 	// If no action selected or action not found, validate inputs from all prompts
-	allInputs := make([]common.Input, 0)
-	for _, prompt := range n.prompts {
-		allInputs = append(allInputs, prompt.Inputs...)
-	}
-
-	return !n.appendMissingInputs(ctx, nodeResp, allInputs)
+	return !n.appendMissingInputs(ctx, nodeResp, n.getAllInputs())
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/core/prompt_node.go:185]: This is not a valid use case. Will revisit this later

File: backend/internal/consent/default_client.go [hunk 10]
```diff
@@ -733,10 +769,18 @@ func (c *defaultClient) buildConsentSearchURL(filter *ConsentSearchFilter) (
...
+		}
+		params.Set("consentTypes", strings.Join(consentTypeStrs, ","))
 	}
 	if len(filter.ConsentStatuses) > 0 {
-		params.Set("consentStatuses", strings.Join(filter.ConsentStatuses, ","))
+		consentStatusStrs := make([]string, 0, len(filter.ConsentStatuses))
+		for _, cs := range filter.ConsentStatuses {
+			consentStatusStrs = append(consentStatusStrs, string(cs))
+		}
+		params.Set("consentStatuses", strings.Join(consentStatusStrs, ","))
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/consent/default_client.go:783]: Will address this separately. @coderabbitai create an issue to track this

File: backend/internal/flow/common/constants.go [hunk 11]
```diff
@@ -152,6 +156,12 @@ const (
 	RuntimeKeyRequiredAttributes = "required_attributes"
 	// RuntimeKeyRequiredLocales holds the space-separated locales requested for claims.
 	RuntimeKeyRequiredLocales = "required_locales"
+	// RuntimeKeyConsentID holds the consent record ID after consent has been recorded.
+	RuntimeKeyConsentID = "consent_id"
+	// RuntimeKeyConsentExpiresAt holds the expiry timestamp of the consent prompt.
+	RuntimeKeyConsentExpiresAt = "consentExpiresAt"
+	// RuntimeKeyConsentedAttributes holds a space-separated set of attributes that the user has consented to share.
+	RuntimeKeyConsentedAttributes = "consented_attributes"
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/common/constants.go:164]: This is fixed

File: backend/internal/flow/executor/constants.go [hunk 12]
```diff
@@ -59,13 +60,14 @@ const (
...
-	userInputInviteToken = "inviteToken"
-	userInputOTP         = "otp"
+	userInputCode             = "code"
+	userInputNonce            = "nonce"
+	userInputOuName           = "ouName"
+	userInputOuHandle         = "ouHandle"
+	userInputOuDesc           = "ouDescription"
+	userInputInviteToken      = "inviteToken"
+	userInputOTP              = "otp"
+	userInputConsentDecisions = "consent_decisions"
```
[the remark this answers is not in the document]
Comment (ThaminduDilshan) [reply backend/internal/flow/executor/constants.go:70]: This is fixed

File: backend/internal/authn/consent/service.go (hunk 3 above)
Comment (ThaminduDilshan) [reply backend/internal/authn/consent/service.go]: This is fixed
